WooCommerce/Classes/Authentication/SiteCredentialLoginUseCase.swift

```diff
@@ -0,0 +1,130 @@
+import enum Alamofire.AFError
+import struct Networking.CookieNonceAuthenticatorConfiguration
+import class Networking.WordPressOrgNetwork
+import Yosemite
+
+protocol SiteCredentialLoginProtocol {
+    func setupHandlers(onLoginSuccess: @escaping () -> Void,
+                       onLoginFailure: @escaping (SiteCredentialLoginError) -> Void)
+
+    func handleLogin(username: String, password: String)
+}
+
+enum SiteCredentialLoginError: Error {
+    static let errorDomain = "SiteCredentialLogin"
+    case wrongCredentials
+    case genericFailure(underlyingError: Error)
+
+    /// Used for tracking error code
+    ///
+    var underlyingError: NSError {
+        switch self {
+        case .wrongCredentials:
+            return NSError(domain: Self.errorDomain, code: 401, userInfo: nil)
+        case .genericFailure(let underlyingError):
+            return underlyingError as NSError
+        }
+    }
+}
+
+/// This use case handles site credential login without the need to use XMLRPC API.
+/// Steps for login:
+/// - Handle cookie authentication with provided credentials.
+/// - Attempt retrieving plugin details. If the request fails with 401 error, the authentication fails.
+///
+final class SiteCredentialLoginUseCase: SiteCredentialLoginProtocol {
+    private let siteURL: String
+    private let stores: StoresManager
+    private var successHandler: (() -> Void)?
+    private var errorHandler: ((SiteCredentialLoginError) -> Void)?
+
+    init(siteURL: String,
+         stores: StoresManager = ServiceLocator.stores) {
+        self.siteURL = siteURL
+        self.stores = stores
+    }
+
+    func setupHandlers(onLoginSuccess: @escaping () -> Void,
+                       onLoginFailure: @escaping (SiteCredentialLoginError) -> Void) {
+        self.successHandler = onLoginSuccess
+        self.errorHandler = onLoginFailure
+    }
+
+    func handleLogin(username: String, password: String) {
+        // Old cookies can make the login succeeds even with incorrect credentials
+        // So we need to clear all cookies before login.
+        clearAllCookies()
+        loginAndAttemptFetchingJetpackPluginDetails(username: username, password: password)
+    }
+}
+
+private extension SiteCredentialLoginUseCase {
+    func clearAllCookies() {
+        let cookieJar = HTTPCookieStorage.shared
```

Comment:
Nit: I think we might be able to unit test that cookies are cleared if we inject HTTPCookieStorage.

Reply:
Added test in cd774d85943a3d595b3cfb9cb599efff1d567900 and a2e59183a3065e6dfa604d8a6c09440e9bcfd5e8.